File: backend/app/api/schemas/survey.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class SurveyReading(BaseModel):
    floor_id: str
    x_norm: float = Field(ge=0, le=1)
    y_norm: float = Field(ge=0, le=1)
    ssid: str
    bssid: str
    band: Literal["2.4GHz", "5GHz", "6GHz"]
    channel: int
    signal_dbm: int
    captured_at: datetime


class SurveySession(BaseModel):
    id: str
    floor_id: str
    captured_at: datetime
    source_filename: str
    location_x_norm: float | None = Field(default=None, ge=0, le=1)
    location_y_norm: float | None = Field(default=None, ge=0, le=1)
    metadata: dict = Field(default_factory=dict)
    networks_excerpt: str | None = None
    interface_excerpt: str | None = None
    log_files: list[str] = Field(default_factory=list)


class SurveyUploadResponse(BaseModel):
    session: SurveySession
    reading_count: int

File: frontend/src/types/survey.ts
export interface Floor {
  id: string;
  name: string;
  map_image_url?: string | null;
}

export interface SurveyReading {
  floor_id: string;
  x_norm: number;
  y_norm: number;
  ssid: string;
  bssid: string;
  band: "2.4GHz" | "5GHz" | "6GHz";
  channel: number;
  signal_dbm: number;
  captured_at: string;
}

export type RssiLevel = "excellent" | "good" | "minimum" | "weak" | "unstable";

export interface HeatmapPoint {
  x_norm: number;
  y_norm: number;
  signal_dbm: number;
  level: RssiLevel;
  band: WifiBand;
}

export type WifiBand = "2.4GHz" | "5GHz" | "6GHz";

export interface BandSummary {
  band: WifiBand;
  count: number;
}

export interface APBandSummaryResponse {
  floor_id: string;
  by_band: BandSummary[];
}

export interface SurveySession {
  id: string;
  floor_id: string;
  captured_at: string;
  source_filename: string;
  location_x_norm?: number | null;
  location_y_norm?: number | null;
  metadata: Record<string, unknown>;
  networks_excerpt?: string | null;
  interface_excerpt?: string | null;
  log_files: string[];
}

export interface SurveyUploadResponse {
  session: SurveySession;
  reading_count: number;
}

export type IssueType = "connection_unavailable" | "latency" | "disconnect" | "auth_failure";
export type IssueSeverity = "low" | "medium" | "high";

export interface IssueReport {
  id: string;
  floor_id: string;
  x_norm: number;
  y_norm: number;
  occurred_at: string;
  ssid: string;
  issue_type: IssueType;
  severity: IssueSeverity;
}

export interface IssueReportCreate {
  floor_id: string;
  x_norm: number;
  y_norm: number;
  occurred_at: string;
  ssid: string;
  issue_type: IssueType;
  severity: IssueSeverity;
}
